# Preprocessing

In [1]:
# Importing libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import datetime

from scipy import stats
from scipy.stats import skew
from scipy.stats import norm
from scipy.stats.stats import pearsonr

%matplotlib inline
%config InlineBackend.figure_format = 'retina'

# Option to display all the dataframe columns
pd.options.display.max_columns = None

In [2]:
# Importing dataset
df = pd.read_csv("dataset/train_original.csv")

# Transform dates from '%d/%m/%Y' to datetime objects.
def transform_date(x):
    date = datetime.datetime.strptime(x, '%d/%m/%Y')
    return date

df['Date'] = df['Date'].map(transform_date)

In [3]:
mi0 =  pd.read_csv("dataset/train_original.csv")

## Overview

In [4]:
df.describe()

StoreID      IsHoliday         IsOpen  HasPromotions  \
count  523021.000000  523021.000000  523021.000000  523021.000000   
mean     1373.953914       0.029718       0.829714       0.383292   
std       216.395820       0.169808       0.375884       0.486189   
min      1000.000000       0.000000       0.000000       0.000000   
25%      1187.000000       0.000000       1.000000       0.000000   
50%      1373.000000       0.000000       1.000000       0.000000   
75%      1561.000000       0.000000       1.000000       1.000000   
max      1748.000000       1.000000       1.000000       1.000000   

       NearestCompetitor         Region  NumberOfCustomers  NumberOfSales  \
count       523021.00000  523021.000000      523021.000000  523021.000000   
mean          8002.71122       5.713151         259.325073    4057.377536   
std          11537.70753       3.357190         185.857102    2729.630381   
min             47.00000       0.000000           0.000000       0.000000   
25%           1057.00000       3.000000         166.000000    2573.000000   
50%           3321.00000       6.000000         251.000000    4020.000000   
75%           9673.00000       9.000000         345.000000    5538.000000   
max          85070.00000      10.000000        2206.000000   26641.000000   

       Region_AreaKM2     Region_GDP  Region_PopulationK     CloudCover  \
count   523021.000000  523021.000000       523021.000000  481840.000000   
mean     12381.265718   14050.547848         3861.519300       5.568627   
std       9140.692427    2755.425404         2816.719484       1.715277   
min        344.000000    9893.000000          816.000000       0.000000   
25%       7215.000000   11849.000000         1293.000000       5.000000   
50%       9643.000000   15017.000000         2770.000000       6.000000   
75%      15566.000000   15931.000000         5727.000000       7.000000   
max      32221.000000   23931.000000         8146.000000       8.000000   

       Max_Dew_PointC  Max_Gust_SpeedKm_h   Max_Humidity  \
count   523021.000000       113074.000000  523021.000000   
mean         8.512014           47.980782      93.523316   
std          5.988302           12.771727       7.617389   
min        -14.000000           21.000000      44.000000   
25%          4.000000           39.000000      90.000000   
50%          9.000000           47.000000      94.000000   
75%         13.000000           55.000000     100.000000   
max         23.000000          106.000000     100.000000   

       Max_Sea_Level_PressurehPa  Max_TemperatureC  Max_VisibilityKm  \
count              523021.000000     523021.000000     511683.000000   
mean                 1018.133184         14.424012         23.560275   
std                     7.985673          8.196295          9.169704   
min                   976.000000        -11.000000          0.000000   
25%                  1013.000000          8.000000         11.000000   
50%                  1018.000000         15.000000         31.000000   
75%                  1023.000000         21.000000         31.000000   
max                  1043.000000         37.000000         31.000000   

       Max_Wind_SpeedKm_h  Mean_Dew_PointC  Mean_Humidity  \
count       523021.000000    523021.000000  523021.000000   
mean            22.223437         6.229677      74.862845   
std              8.725281         5.881529      12.939739   
min              3.000000       -15.000000      30.000000   
25%             14.000000         2.000000      66.000000   
50%             21.000000         7.000000      76.000000   
75%             26.000000        11.000000      85.000000   
max            101.000000        20.000000     100.000000   

       Mean_Sea_Level_PressurehPa  Mean_TemperatureC  Mean_VisibilityKm  \
count               523021.000000      523021.000000      511683.000000   
mean                  1015.002422          10.273205          12.151025   
std                      8.261523           7.041167    

The dimension of the dataframe

In [5]:
n_instances, n_features = df.shape

So, there are **36** features and **523021** instances

### Given features

The features we are initially given are: 


 * **StoreID** : Store ID,
 * **Date** : Observation Date,
 * **IsHoliday** : Regional Holiday,
 * **IsOpen** : Store Open,
 * **HasPromotions** : The Store is currently running promotions,
 * **StoreType** : Type of Store (Standard Market, Super Market, Hyper Market, Shopping Centre),
 * **AssortmentType** : Type of Products (General, With Fish Department, With Non-Food Department),
 * **NearestCompetitor** : Distance (meters) to the nearest competitor store,
 * **Region** : Region ID ,
 * **NumberOfCustomers** : Number of Customer,
 * **NumberOfSales** :Number of Products Sold **(Target Value)**,
 * **Region_AreaKM2** : Region Size in which the Store is located (Km<sup>2</sup>),
 * **Region_GDP** : Region Gross Domestic Product,
 * **Region_PopulationK** : Region Population Size (k),
 * **CloudCover** : Level of Covering(0-9),
 * **Events** : Meteorological event,
 * **Max_Dew_PointC** : Max Dew Point (Celsius),
 * **Max_Gust_SpeedKm_h** : Max Gust Speed (Km/h),
 * **Max_Humidity** : Max Humidity, 
 * **Max_Sea_Level_PressurehPa** : Max Pressure (hPa),
 * **Max_TemperatureC** : Max Temperature (Celsius),
 * **Max_VisibilityKm** : Max Visibility (Km),
 * **Max_Wind_SpeedKm_h** : Max Wind Speed (Km/h),
 * **Mean_Dew_PointC** : Mean Dew Point (Celsius),
 * **Mean_Humidity** : Mean Humidity,
 * **Mean_Sea_Level_PressurehPa** : Mean Pressure (hPa),
 * **Mean_TemperatureC** : Mean Temperature (Celsius),
 * **Mean_VisibilityKm** : Mean Visibility (Km),
 * **Mean_Wind_SpeedKm_h** : Mean Window Speed (Km/h),
 * **Min_Dew_PointC** : Min Dew Point (Celsius),
 * **Min_Humidity** : Min Humidity,
 * **Min_Sea_Level_PressurehPa** : Min Pressure (hPa),
 * **Min_TemperatureC** : Min Temperature (Celsius),
 * **Min_VisibilitykM** : Min Visibility (Km),
 * **Precipitationmm** : Precipitation Amount (mm),
 * **WindDirDegrees** : Window Direction (degrees).
 
 

In [6]:
categorical_variables_mask = df.dtypes==object
categorical_variables = df.columns[categorical_variables_mask]
numerical_variables = df.columns[~categorical_variables_mask]
print('{} Numerical Variables: '.format(len(numerical_variables)), numerical_variables)
print('{} Categorical Variables: '.format(len(categorical_variables)), categorical_variables)

33 Numerical Variables:  Index(['StoreID', 'Date', 'IsHoliday', 'IsOpen', 'HasPromotions',
       'NearestCompetitor', 'Region', 'NumberOfCustomers', 'NumberOfSales',
       'Region_AreaKM2', 'Region_GDP', 'Region_PopulationK', 'CloudCover',
       'Max_Dew_PointC', 'Max_Gust_SpeedKm_h', 'Max_Humidity',
       'Max_Sea_Level_PressurehPa', 'Max_TemperatureC', 'Max_VisibilityKm',
       'Max_Wind_SpeedKm_h', 'Mean_Dew_PointC', 'Mean_Humidity',
       'Mean_Sea_Level_PressurehPa', 'Mean_TemperatureC', 'Mean_VisibilityKm',
       'Mean_Wind_SpeedKm_h', 'Min_Dew_PointC', 'Min_Humidity',
       'Min_Sea_Level_PressurehPa', 'Min_TemperatureC', 'Min_VisibilitykM',
       'Precipitationmm', 'WindDirDegrees'],
      dtype='object')
3 Categorical Variables:  Index(['StoreType', 'AssortmentType', 'Events'], dtype='object')


We can see that there are 4 categorical variables:

* StoreType, with 4 levels (Standard Market, Super Market, Hyper Market, Shopping Centre)
* AssortmentType, with 3 levels (General, With Fish Department, With Non-Food Department)
* Events, with 22 levels (NaN, and combinations of 5 atmospherical events: Rain, Fog, Hail, Snow, Thunderstorm)
* Date, with 729 levels (every day from March 1, 2016 to February 28, 2018)

So we have to find ways to convert these categorical variables to numerical.

### Missing values

The numbers of null values for each columns is:

In [7]:
#Only features with missing values, the others are omitted
missing_values = round(df.isnull().sum() / n_instances * 100, 3)
features_with_missings = missing_values.drop(missing_values[missing_values == 0].index).sort_values(ascending=False)
features_with_missings

Max_Gust_SpeedKm_h    78.381
Events                23.727
CloudCover             7.874
Min_VisibilitykM       2.168
Mean_VisibilityKm      2.168
Max_VisibilityKm       2.168
dtype: float64

## Data cleaning and feature engineering

After the initial data exploration, we'll clean the dataset and try to engineer some new interesting features.

First of all, we _drop_ the **Region ID**. It makes no sense as a numerical variable because we don't want it to act as an ordering between the different regions, and there are already 3 features which describe the most interesting things about a region:

* Region_AreaKM2
* Region_GDP
* Region_PopulationK

(the region ID will be dropped before the training for the reason just said, but we need it for the computation of the error)

df = df.drop(labels='Region', axis=1)

Next, we convert **StoreType** to a numerical variable by assigning a fixed map of values. We believe this makes sense because there is an order between the 4 levels of the original categorical variable: e.g. 'Hyper Market' is _bigger_ than 'Super Market. So, in a sense, this feature would be an indicator of how _big_ the store is.

In [14]:
df["StoreType"] = df["StoreType"].map({'Standard Market': 0, 'Super Market': 1, 'Hyper Market': 2, 'Shopping Center': 3})

For **AssortmentType**, we use the pandas function get_dummies, which converts the categorical variable into dummy/indicator variables. In this case, it doesn't make sense to use a single numerical feature with arbitrary numbers because the type of products a store has is better modeled as a boolean yes/no feature (e.g. has / hasn't got fish) rather than with a single ordered variable.

In [15]:
df = pd.get_dummies(df, columns=['AssortmentType'])

We can also drop the generated 'AssortmentType_General' feature because it is a combination of the other two ('AssortmentType_With Non-Food Department' and 'AssortmentType_With Fish Department')

In [16]:
df = df.drop(labels=['AssortmentType_General'], axis=1)

The **Events** feature has many levels (21, not including NaN) and also many missing values (nearly 24%).

We believe it can safely be assumed that a NaN in this feature simply means that none of the events (Rain, Fog, Hail, Snow, Thunderstorm) occurred. This is because when Events is NaN the other features suggest a better weather (e.g. higher temperature and visibility) than when Events is not NaN.

We convert this feature in 5 boolean features that are active when the corresponding atmospherical event happened:

In [17]:
for e in ['Rain', 'Snow', 'Fog', 'Hail', 'Thunderstorm']:
    df[e] = df.loc[:, 'Events'].astype('str').apply(lambda x: int(e in x))
    
df = df.drop(labels='Events', axis=1)

We add a couple of features signaling if it is a Saturday or a Sunday:

In [18]:
def is_saturday(day):
    return int(day.weekday() == 5)

def is_sunday(day):
    return int(day.weekday() == 6)

df['IsSaturday'] = df['Date'].map(is_saturday)
df['IsSunday'] = df['Date'].map(is_sunday)

Now we have to **deal with the missing values**.

The approach that we are going to use is:

* see which feature has the highest correlation with the feature we want to fix;
* use that feature as a guide to fill in the missing values.

We are now going to define a handy function to do so:

In [19]:
def fill_missing(df, feature, predictor):
    feature_not_null        = df.dropna(subset=[feature])
    
    feature_mean            = feature_not_null[feature].mean()
    predictor_mean          = feature_not_null[predictor].mean()
    feature_predictor_ratio = feature_mean / predictor_mean
    
    inverse_correlation = df.corr().loc[[feature], [predictor]].iloc[0, 0] < 0
    
    def f(x):
        current_feature_value   = x[feature]
        current_predictor_value = x[predictor]
        
        new_feature_value = 0.0
        
        if np.isnan(current_feature_value):
            if inverse_correlation:
                new_feature_value = feature_mean - ((current_predictor_value - predictor_mean) * feature_predictor_ratio)
            else:
                new_feature_value = feature_mean + ((current_predictor_value - predictor_mean) * feature_predictor_ratio)
        else:
            new_feature_value = current_feature_value
            
        return new_feature_value
    
    df[feature] = df.loc[:, [feature, predictor]].apply(f, axis=1)
    
    return df

This function also deals with _inverse correlation_, that is when a variable increases then the other one decreases.

The feature with the most NaNs is **Max_Gust_SpeedKm_h** with little more than 78%. Obviously we can't drop 78% of the instances, but we don't want to drop the feature either.

We notice that, not surprisingly, this feature is very highly correlated to **Max_Wind_SpeedKm_h** (0.879945):

In [20]:
df.corr().loc[['Max_Gust_SpeedKm_h'], ['Max_Wind_SpeedKm_h']]

Max_Wind_SpeedKm_h
Max_Gust_SpeedKm_h            0.879945

So we use that feature as a guide to fill in the missing values of Max_Gust_SpeedKm_h:

In [21]:
df = fill_missing(df, 'Max_Gust_SpeedKm_h', 'Max_Wind_SpeedKm_h')

The same can be said for **CloudCover**. The most correlated feature is Min_Humidity:

In [22]:
df.corr().loc[['CloudCover'], ['Min_Humidity']]

Min_Humidity
CloudCover      0.419168

So we use that feature as a guide to fill in the missing values of CloudCover:

In [23]:
df = fill_missing(df, 'CloudCover', 'Min_Humidity')

Now we want to add features specifying the number of sales for the previous day, and the _mean_ number of sales for the previous 3 days, week, and month. Obviously this data is related to a single store, so first of all we have to split the dataset in chunks, based on the _StoreID_.

We also add a couple of boolean features signaling if the store was open the previous day, and if the store will be open the following day.

In [24]:
def ordinal_date(x):
    d = x['Date']
    date = datetime.date(d.year, d.month, d.day)
    return datetime.date.toordinal(date)

df['OrdinalDate'] = df.loc[:, ['Date']].apply(ordinal_date, axis=1)

base_date = min(df['OrdinalDate'])

df['OrdinalDate'] = df['OrdinalDate'].apply(lambda x: x - base_date)

In [25]:
# Divide dataset in chunks based on the Store ID
stores = []
for store in df['StoreID'].unique():
    stores.append( df.loc[(df['StoreID'] == store)] )

In [26]:
def was_open_yesterday(x):
    date = x['OrdinalDate']
    store_index = x['StoreID'] - 1000
    if date < 30:
        return 1
    that_date = stores[store_index].loc[(stores[store_index]['OrdinalDate'] == date - 1)]
    if len(that_date) < 1:
        return 0
    else:
        return that_date['IsOpen'].item()
    
df['WasOpenYesterday'] = df.loc[:, ['OrdinalDate', 'StoreID']].apply(was_open_yesterday, axis=1)

In [27]:
def is_open_tomorrow(x):
    date = x['OrdinalDate']
    store_index = x['StoreID'] - 1000
    if date == 729:
        return 1
    that_date = stores[store_index].loc[(stores[store_index]['OrdinalDate'] == date + 1)]
    if len(that_date) < 1:
        return 0
    else:
        return that_date['IsOpen'].item()

df['IsOpenTomorrow'] = df.loc[:, ['OrdinalDate', 'StoreID']].apply(is_open_tomorrow, axis=1)

In [28]:
df.corr().loc[['NumberOfSales'], ['WasOpenYesterday', 'IsOpenTomorrow']]

WasOpenYesterday  IsOpenTomorrow
NumberOfSales          -0.20478        0.021366

def find_ysales(x):
    date, sales, store_index = x['OrdinalDate'], x['NumberOfSales'], x['StoreID']-1000
    if date < 30:
        return 0
    else:
        that_date = stores[store_index].loc[(stores[store_index]['OrdinalDate'] == date - 1)]
        if len(that_date) < 1:
            return 0
        else:
            return that_date['NumberOfSales'].item()

df['YesterdaySales'] = df.loc[:, ['OrdinalDate', 'NumberOfSales', 'StoreID']].apply(find_ysales, axis=1)

def find_3days_sales(x):
    date, sales, store_index = x['OrdinalDate'], x['NumberOfSales'], x['StoreID']-1000
    if date < 30:
        return 0
    else:
        that_date = stores[store_index].loc[(stores[store_index]['OrdinalDate'] <= date - 1) & (stores[store_index]['OrdinalDate'] > date - 4)]
        if len(that_date) < 1:
            return 0
        else:
            return that_date['NumberOfSales'].mean()

df['Previous3DaysSales'] = df.loc[:, ['OrdinalDate', 'NumberOfSales', 'StoreID']].apply(find_3days_sales, axis=1)

def find_7days_sales(x):
    date, sales, store_index = x['OrdinalDate'], x['NumberOfSales'], x['StoreID']-1000
    if date < 30:
        return 0
    else:
        that_date = stores[store_index].loc[(stores[store_index]['OrdinalDate'] <= date - 1) & (stores[store_index]['OrdinalDate'] > date - 8)]
        if len(that_date) < 1:
            return 0
        else:
            return that_date['NumberOfSales'].mean()

df['PreviousWeekSales'] = df.loc[:, ['OrdinalDate', 'NumberOfSales', 'StoreID']].apply(find_7days_sales, axis=1)

def find_30days_sales(x):
    date, sales, store_index = x['OrdinalDate'], x['NumberOfSales'], x['StoreID']-1000
    if date < 30:
        return 0
    else:
        that_date = stores[store_index].loc[(stores[store_index]['OrdinalDate'] <= date - 1) & (stores[store_index]['OrdinalDate'] > date - 31)]
        if len(that_date) < 1:
            return 0
        else:
            return that_date['NumberOfSales'].mean()
        
df['PreviousMonthSales'] = df.loc[:, ['OrdinalDate', 'NumberOfSales', 'StoreID']].apply(find_30days_sales, axis=1)

The 3 visibility features are the only ones left with a few missing values. Let's take care of them in the same way we did before.

First, we find the features which are most correlated to them:

In [29]:
df.corr().loc[['Min_VisibilitykM', 'Mean_VisibilityKm', 'Max_VisibilityKm'], ['Mean_Humidity', 'Region_AreaKM2']]

Mean_Humidity  Region_AreaKM2
Min_VisibilitykM       -0.620345       -0.047109
Mean_VisibilityKm      -0.603040       -0.220113
Max_VisibilityKm       -0.259888       -0.443495

And then we use them for filling in the missing values:

In [30]:
df = fill_missing(df, 'Min_VisibilitykM', 'Mean_Humidity')
df = fill_missing(df, 'Mean_VisibilityKm', 'Mean_Humidity')
df = fill_missing(df, 'Max_VisibilityKm', 'Region_AreaKM2')

In [31]:
df.corr().loc[['Min_VisibilitykM', 'Mean_VisibilityKm', 'Max_VisibilityKm'], ['Mean_Humidity', 'Region_AreaKM2']]

Mean_Humidity  Region_AreaKM2
Min_VisibilitykM       -0.615748       -0.045687
Mean_VisibilityKm      -0.606757       -0.214194
Max_VisibilityKm       -0.233406       -0.474210

Let's remove the rows where the store is closed (IsOpen = 0) and the IsOpen feature itself:

In [32]:
df = df.drop(df[df.IsOpen == 0].index)
df = df.drop(labels=['IsOpen'], axis=1)

Let's add a feature for the mean **NumberOfSales** for all the stores in that month:

In [33]:
df['YearAndMonth'] = df.Date.map(lambda d: d.strftime('%Y-%m'))
mean_month_sales = df.groupby('YearAndMonth').mean().NumberOfSales.to_dict()
df['MeanMonthSales'] = df.Date.map(lambda d: mean_month_sales[d.strftime('%Y-%m')])
df = df.drop(labels=['YearAndMonth'], axis=1)

Also, let's add a feature for the mean **NumberOfSales** for that store:

In [34]:
mean_store_sales = dict()

for store_id in df.StoreID.unique():
    rows = df[df.StoreID == store_id]
    mean_store_sales[store_id] = rows.NumberOfSales.mean()
    
df['MeanStoreSales'] = df.StoreID.map(mean_store_sales)

The preprocessing is now finished. The only variable which is still categorical is 'Date', it needs to be either dropped or converted to some numerical variable(s). Here are some final statistics. As we can see, there are no more missing values.

In [35]:
n_instances, n_features = df.shape

print('Number of instances: {}\nNumber of features: {}'.format(n_instances, n_features))

#Only features with missing values, the others are omitted
missing_values = round(df.isnull().sum() / n_instances * 100, 3)
features_with_missings = missing_values.drop(missing_values[missing_values == 0].index).sort_values(ascending=False)
print(features_with_missings)

df.describe()

Number of instances: 433958
Number of features: 47
Series([], dtype: float64)


StoreID      IsHoliday  HasPromotions      StoreType  \
count  433958.000000  433958.000000  433958.000000  433958.000000   
mean     1373.938579       0.000945       0.449507       1.304594   
std       216.245290       0.030723       0.497444       0.899660   
min      1000.000000       0.000000       0.000000       0.000000   
25%      1188.000000       0.000000       0.000000       0.000000   
50%      1373.000000       0.000000       0.000000       2.000000   
75%      1561.000000       0.000000       1.000000       2.000000   
max      1748.000000       1.000000       1.000000       3.000000   

       NearestCompetitor         Region  NumberOfCustomers  NumberOfSales  \
count      433958.000000  433958.000000      433958.000000  433958.000000   
mean         8042.993882       5.727451         312.547433    4890.089953   
std         11640.203554       3.357071         158.106619    2215.403562   
min            47.000000       0.000000           0.000000       0.000000   
25%          1057.000000       3.000000         214.000000    3386.000000   
50%          3307.000000       6.000000         278.000000    4469.500000   
75%          9761.000000       9.000000         369.000000    5911.000000   
max         85070.000000      10.000000        2206.000000   26641.000000   

       Region_AreaKM2     Region_GDP  Region_PopulationK     CloudCover  \
count   433958.000000  433958.000000       433958.000000  433958.000000   
mean     12313.341916   14045.556837         3845.765475       5.397822   
std       9114.075588    2760.416990         2816.324768       1.822572   
min        344.000000    9893.000000          816.000000       0.000000   
25%       7215.000000   11849.000000         1293.000000       4.000000   
50%       9337.000000   15017.000000         1892.000000       6.000000   
75%      15566.000000   15931.000000         5727.000000       7.000000   
max      32221.000000   23931.000000         8146.000000      10.277723   

       Max_Dew_PointC  Max_Gust_SpeedKm_h   Max_Humidity  \
count   433958.000000       433958.000000  433958.000000   
mean         8.566239           33.887436      93.534727   
std          5.987954           13.012629       7.550378   
min        -14.000000            4.590606      44.000000   
25%          4.000000           24.483234      89.000000   
50%          9.000000           32.134244      94.000000   
75%         13.000000           40.000000     100.000000   
max         23.000000          154.550413     100.000000   

       Max_Sea_Level_PressurehPa  Max_TemperatureC  Max_VisibilityKm  \
count              433958.000000     433958.000000     433958.000000   
mean                 1018.144102         14.511649         23.348092   
std                     7.974594          8.220396          9.666694   
min                   976.000000        -11.000000        -14.588085   
25%                  1013.000000          8.000000         11.000000   
50%                  1018.000000         15.000000         31.000000   
75%                  1023.000000         21.000000         31.000000   
max                  1043.000000         37.000000         46.461731   

       Max_Wind_SpeedKm_h  Mean_Dew_PointC  Mean_Humidity  \
count       433958.000000    433958.000000  433958.000000   
mean            22.158944         6.285320      74.859115   
std              8.751433         5.862926      12.928407   
min              3.000000       -15.000000      30.000000   
25%             14.000000         2.000000      66.000000   
50%             21.000000         7.000000      76.000000   
75%             26.000000        11.000000      85.000000   
max            101.000000        20.000000     100.000000   

       Mean_Sea_Level_PressurehPa  Mean_TemperatureC  Mean_VisibilityKm  \
count               433958.000000       433958.00000      433958.000000   
mean                  1014.979698           10.33585          12.211131   
std                      8.262854            7.04847    

In [36]:
df = df.drop(['StoreID','NumberOfCustomers'], axis=1)

In [37]:
df_date_label = df.loc[:,['Date', 'NumberOfSales','Region']]
df_no_date = df.drop(['Date', 'NumberOfSales','Region'], axis=1)
df_norm = (df_no_date - df_no_date.min()) / (df_no_date.max() - df_no_date.min())
df = df_norm
df['Date'] = df_date_label['Date']
df['NumberOfSales'] = df_date_label['NumberOfSales']
df['Region'] = df_date_label['Region']

In [38]:
df_val2 = df.loc[(df['Date']>='2018-01-01')]
df_rest = df.loc[(df['Date']<'2018-01-01')]

In [40]:
msk = np.random.rand(len(df_rest)) < 0.9
df_train = df_rest[msk]
df_val1= df_rest[~msk]

In [43]:
print(df_train.shape, df_val1.shape)

(357647, 45) (39400, 45)


df_val2 = df_val2.drop(['Date'], axis=1)
df_val1 = df_val1.drop(['Date'], axis=1)
df_train = df_train.drop(['Date'], axis=1)

In [44]:
df_train.to_csv('train_rand.csv')
df_val1.to_csv('train_val1_rand.csv')
df_val2.to_csv('train_val2.csv')